### PR TITLE
Order Processor

### PR DESCRIPTION
**Pull request by amnmtr**
Adding Order Processing Flow...
Flow starts with rabbitmq consumer as a rake task, It gets order message with status updates. the rake task offloads the Order processing to a sidekiq worker that is responsible to persist the order while updating existing order with the status update ( while not updating the order status if message are out of order )

### CODE REVIEW

File: app/workers/order_processor.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+class OrderProcessor
+  include Sidekiq::Worker
+  sidekiq_options backtrace: true, queue: :deliveries
+  def perform(consumed_order)
+    begin
+      puts 'sidekiq worker processed ' + consumed_order.to_s
+      consumed_order = JSON.parse(consumed_order)
+      order = Order.find_or_initialize_by(order_id: consumed_order['order_id'] )
+      if order.publish_time.nil? || order.publish_time < consumed_order['publish_time']
```
Comment (ronits-wework) [inline app/workers/order_processor.rb:9]: it's a little difficult to understand that order is either created or updated
then in the rest of the flow you check whether order is new by checking that its values are nil
I think it should be separated to 2 different methods - one that creates and one that updates

File: app/workers/order_processor.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    rescue => exception
+      exception.to_s
+    end
+  end
+
+
+  private
+  def fill_extra_info (consumed_order)
+      extra_info = {}
+      extra_info[:courier] = consumed_order['courier'] if !consumed_order['courier'].nil?
```
Comment (ronits-wework) [inline app/workers/order_processor.rb:35]: `!consumed_order['courier'].nil?` can be shortened to `consumed_order['courier'].present?` so it's not a double negative

Comment (ronits-wework) [reply app/workers/order_processor.rb:35]: you could also shorten this to:
```ruby
extra_info_keys = [:courier, :cook, :estimated_time_of_arrival, :signed_by]
extra_info_keys.each do |key|
    extra_info[key] = consumed_order[key.to_s] if consumed_order[key.to_s].present?
end
```

File: app/workers/order_processor.rb (hunk 1 above)
Comment (amnmtr) [reply app/workers/order_processor.rb:9]: i'm always upserting so it doesn't matter new Order or not .. the nil check are for merging purposes as i don't want nil overrides.. same order can arrive multiple times with added attributes and i want to merge them always

File: app/workers/order_processor.rb (hunk 2 above)
Comment (amnmtr) [reply app/workers/order_processor.rb:35]: didn't use present as i understood that it returns false if the value is a boolean false, while in that case i want still to perform an override ... at least thats what was documented .... the loop is cool though 8-)

File: db/migrate/20190806135259_create_orders.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  def up
+    create_table :orders do |t|
+      t.references :restaurant, index: true, foreign_key: true
+      t.string :order_id, unique: true, index: true
+      t.string :customer_name
+      t.datetime :time
+      t.datetime :publish_time
+      t.string :status
+      t.json :info, default: {}
+      #Ex:- :default =>''
```
Comment (ronits-wework) [inline db/migrate/20190806135259_create_orders.rb:11]: remove comment

File: lib/tasks/order_consumer.rake [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    conn.start
+    channel = conn.create_channel
+    queue = channel.queue( 'delivery.status_updated', auto_delete: true)
+
+    puts "subscribing"
+    queue.subscribe(block: true) do |_, _, payload|
+  # DO STUFF
+      puts "Consuming!! : " + payload.to_s
+
+      puts "sending to sidekiq worker : " + payload.to_s
```
Comment (ronits-wework) [inline lib/tasks/order_consumer.rake:17]: why do you print payload.to_s twice?

Comment (amnmtr) [reply lib/tasks/order_consumer.rake:17]: Thought i deleted one of them guess not :)

File: spec/models/order_spec.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    it { should validate_presence_of(:publish_time) }
+    it { should validate_presence_of(:status) }
+    
+    it { should validate_uniqueness_of(:order_id) }
+    
+    it { should belong_to(:restaurant) } 
+
+    
+    it 'ensure status valid value' do
+      order = Order.create(order_id: 'order_uuid', customer_name: 'name dummy', time: '2019-08-06 18:29:57', publish_time: '2019-08-06 18:29:57', status: 'DummyStatus')
```
Comment (ronits-wework) [inline spec/models/order_spec.rb:34]: 1. Create a factory for order that will automatically generate dummy data
2. Use `build` instead of `create` - we don't need to persist to the DB to check validity (I think)
3. Just so you know, when there's a boolean method ending with a question mark, rspec has magic syntax:
`expect(order).to_not be_valid`

File: app/workers/order_processor.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
+class OrderProcessor
+  include Sidekiq::Worker
+  sidekiq_options backtrace: true, queue: :deliveries
+  def perform(consumed_order)
+    begin
```
Comment (ronits-wework) [inline app/workers/order_processor.rb:5]: you don't have to use begin to rescue in a method
you can just do:
```ruby
def perform(consumed_order)
    # code
rescue StandardError => e
   # rescue code
end
```

File: app/workers/order_processor.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+      if(order.info.nil?)
+        order.info = extra_info
+      else
+        order.info.merge(extra_info)
+      end
+
+      order.save!
+    rescue => exception
+      exception.to_s
```
Comment (ronits-wework) [inline app/workers/order_processor.rb:27]: what does this do? did you mean to log it?
